Update bindgen

hannesweisbach opened the pull request:
Updating for a solution for issue rust-lang/rust-bindgen#2312 of rust-bindgen.

I'm in the process to generate bindings for softdevice 332, when I encountered the error

```
+ for s in s332
+ ./nrf-softdevice-gen/target/release/nrf-softdevice-gen ./softdevice/s332/headers ./nrf-softdevice-s332/src/bindings.rs
thread 'main' panicked at /home/hannesweisbach/.cargo/registry/src/index.crates.io-6f17d22bba15001f/bindgen-0.55.1/src/ir/context.rs:838:9:
"ble_gap_evt_timeout_t_union_(unnamed_at__/tmp/ble_gap_h_1196_3)" is not a valid Ident
stack backtrace:
```

This SO post  https://stackoverflow.com/questions/76443280/rust-bindgen-causes-a-is-not-a-valid-ident-error-on-build pointed me to rust-lang/rust-bindgen#2312.

The version bump solves the error above.

hannesweisbach commented:
Edit: changed from 0.71.1 down to 0.70.0, because 0.71.1 would generate stray "unsafe" keywords. Not sure what's up. 0.70.0 seems to work though.

alexmoon commented:
It looks like bindings generated with the new version of bindgen are not backwards-compatible. If you want to update bindgen, please run `gen.sh` so that the repository remains consistent and reproducible and make sure that all tests pass (by running `ci.sh`). Ideally you can get it to work by just adjusting the bindgen configuration.

hannesweisbach commented:
Thanks for your feedback!

So not quite as straight-forward as I hoped. I want to mention that I get these types of warnings when I run `gen.sh`:

```bash
+ ./nrf-softdevice-gen/target/release/nrf-softdevice-gen ./softdevice/s140/headers ./nrf-softdevice-s140/src/bindings.rs
Warning: can't set `imports_granularity = Module`, unstable features are only available in nightly channel.
Warning: can't set `group_imports = StdExternalCrate`, unstable features are only available in nightly channel.
Warning: can't set `imports_granularity = Module`, unstable features are only available in nightly channel.
Warning: can't set `group_imports = StdExternalCrate`, unstable features are only available in nightly channel.
+ rustfmt ./nrf-softdevice-s140/src/bindings.rs
Warning: can't set `imports_granularity = Module`, unstable features are only available in nightly channel.
Warning: can't set `group_imports = StdExternalCrate`, unstable features are only available in nightly channel.
+ cd nrf-softdevice-s140
+ cargo build --target thumbv7em-none-eabihf
```
Do you know if they are expected/benign?

The new generated bindings do not have as nice formatted docs as the old ones. Instead of multiple lines there is just one `#[doc]`macro with embedded newlines. There is also an issue open rust-lang/rust-bindgen#3073 regarding this issue. As a stop-gap there is a set of `rustfmt` options to re-format the comments in multiline but with `///` instead of `#[doc]`. Would that be acceptable?

The new `bind-gen` version introduced `_bitfield_align_1`-members in structs which required to be initialized manually, so I added initializers in `nrf-softdevice/` and `examples/`. `./ci.sh` executes now without errors. I'm still investigating whether there is a setting for `bindgen` to control generation of these members.

alexmoon commented:
> Do you know if they are expected/benign?

Yes, those warnings are fine. We use some rustfmt options that are only on nightly.

> The new generated bindings do not have as nice formatted docs as the old ones. Instead of multiple lines there is just one `#[doc]`macro with embedded newlines. There is also an issue open [rust-lang/rust-bindgen#3073](https://github.com/rust-lang/rust-bindgen/issues/3073) regarding this issue. As a stop-gap there is a set of `rustfmt` options to re-format the comments in multiline but with `///` instead of `#[doc]`. Would that be acceptable?

`#[doc]` macros with embedded newlines are exactly identical to multiple `#[doc]` macros as far as rustdoc is concerned, so that's fine.

> The new `bindgen` version introduced `_bitfield_align_1`-members in structs which required to be initialized manually, so I added initializers in `nrf-softdevice/` and `examples/`. `./ci.sh` executes now without errors. I'm still investigating whether there is a setting for `bindgen` to control generation of these members.

This is more of a problem, because it breaks existing code. In most cases it's also useless because the bitfield is byte aligned anyway. Let me know what you find.

hannesweisbach commented:
Sorry for the long silence.

> Edit: changed from 0.71.1 down to 0.70.0, because 0.71.1 would generate stray "unsafe" keywords. Not sure what's up. 0.70.0 seems to work though.

I've updated to bindgen 0.72.0, which required a small change in a regexp (add "unsafe" to it) to do the right thing.



> > The new `bindgen` version introduced `_bitfield_align_1`-members in structs which required to be initialized manually, so I added initializers in `nrf-softdevice/` and `examples/`. `./ci.sh` executes now without errors. I'm still investigating whether there is a setting for `bindgen` to control generation of these members.
> 
> This is more of a problem, because it breaks existing code. In most cases it's also useless because the bitfield is byte aligned anyway. Let me know what you find.

I opened a question in rust-lang/rust-bindgen#3170, alas no takers. The change that generates the `_bitfield_align_x`-members was added in Dec 2020, only a couple of months after bindgen 0.55.1 was released (Aug 2020, iirc). The change looks intentional to me.

I also tried to solve the issue using RFC3681 (rust-lang/rust#132162), which implements default field values and seems on a good track to be mainlined, but that would (1) still break the API, (2) require the same amount of changes to fix it, and (3) require nightly compilers. See https://github.com/embassy-rs/nrf-softdevice/commit/0edafd031b2f532d549639b41973f89220bb913b, if you're curious.

At this point I think any update from rust-bindgen 0.55.1 will break the API (at least regarding bitfields). With this in mind, I'd understand if you want to not update bindgen for the time being.
